Add channel name on email notification.

sjstyle opened the pull request:
#### Summary

The email is sent when someone who are in channel mentioned. 
But, there is no information about channel name in email notification. 
(cf. push notification has channel name information on it.)

For better identification, I think that email notification need to provide channel name to mail subject.

#### Ticket Link

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [ ] Added or updated unit tests (required for all new features)
- [ ] Added API documentation (required for all new APIs)
- [ ] All new/modified APIs include changes to the drivers
- [ ] Has enterprise changes (please link)
- [ ] Has UI changes
- [x] Includes text changes and localization files updated
- [ ] Touches critical sections of the codebase (auth, upgrade, etc.)

mattermod commented:
Thanks @sjstyle for the pull request!

 Per the [**CONTRIBUTING.md**](https://github.com/mattermost/platform/blob/master/CONTRIBUTING.md#submitting-a-pull-request) file displayed when you created this pull request, we need to add you to the list of [approved contributors](https://docs.google.com/spreadsheets/d/1NTCeG-iL_VS9bFqtmHSfwETo5f-8MQ7oMDE5IUYJi_Y/pubhtml?gid=0&single=true) for the Mattermost project.

**Please help complete the Mattermost [contribution license agreement](http://www.mattermost.org/mattermost-contributor-agreement/)?**

This is a standard procedure for many open source projects. Your form should be processed within 24 hours and reviewers for your pull request will be able to proceed.

Please let us know if you have any questions.

We are very happy to have you join our growing community!

mattermod commented:
Requested test server will be created if build is successful (all checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created.

Link: http://i-b9efb964.spinmint.com:8065/pr3824

Email: `test@test.com`

Email2: `test2@test.com`

Password: `passwd`

Instance ID: i-b9efb964

mattermod commented:
Spinmint test server destroyed

mattermod commented:
Requested test server will be created if build is successful (all checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created.

Link: http://i-20eabcfd.spinmint.com:8065/pr3824

Email: `test@test.com`

Email2: `test2@test.com`

Password: `passwd`

Instance ID: i-20eabcfd

esethna commented:
Hi @sjstyle, thanks for the PR! In order to take this we would need to make the subject string more readable to fit with our [UX guidelines](https://docs.mattermost.com/developer/fx-guidelines.html). Proposed tweaks:

1. Private Groups and Channels:
`New Mention in [Team_Display_Name] ([Channel_Display_Name]) on [Date]` for example, "New Mention in Contributors (Town Square) on August 18, 2016"
2. Direct Messages:
`New Direct Message in [Team_Display_Name] from [Sender_Name] on [Date]` for example, "New Direct Message in Contributors from Eric Sethna on August 18, 2016"
    - Note that [Sender_Name] would have to respect the **Teammate Name Display** setting in account settings.

mattermod commented:
Spinmint test server destroyed

sjstyle commented:
I modified the code that you suggested. Can you review it again?

esethna commented:
@sjstyle thanks for the tweaks! Sorry, seems there are conflicts on the branch and the test server cannot start. Can you please help resolve?

sjstyle commented:
@esethna I solved conflicts on the branch

mattermod commented:
Requested test server will be created if build is successful (all checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created.

Link: http://i-7c6c33a1.spinmint.com:8065/pr3824

Email: `test@test.com`

Email2: `test2@test.com`

Password: `passwd`

Instance ID: i-7c6c33a1

esethna commented:
Excellent, thank you @sjstyle! +1

mattermod commented:
Spinmint test server destroyed

enahum commented:
Im confused is it the `ChannelName` or the `senderName`? which one should be used here?

hmhealey commented on api/post.go:
```diff
@@ -776,11 +780,34 @@ func sendNotificationEmail(post *model.Post, user *model.User, channel *model.Ch
+		if result := <-Srv.Store.Preference().Get(user.Id, model.PREFERENCE_CATEGORY_DISPLAY_SETTINGS, "name_format"); result.Err != nil {
+			channelName = senderName
+		} else {
+			if result.Data.(model.Preference).Value == "nickname_full_name" {
```
You can actually use `User.GetDisplayNameForPreference(nameFormat)` here instead of needing this whole if-else block.

Can you also add a null check for `sender` here? We shouldn't ever pass a null sender into the function since there's no system messages that are sent over direct channels, but it would be nice to have just in case those are added.

Other than that, everything else looks good. Thanks!

sjstyle commented:
@enahum 
I changed mail parameter templates for more readability.

@hmhealey 
I add null checks and refactored using `GetDisplayNameForPreference` method. Thanks.

enahum commented:
+1

coreyhulen commented:
+1
